src/libraries/System.Linq/src/System/Linq/Count.cs

```diff
@@ -72,6 +72,53 @@ public static int Count<TSource>(this IEnumerable<TSource> source, Func<TSource,
             return count;
         }
 
+        /// <summary>
+        ///   Attempts to determine the number of elements in a sequence without forcing an enumeration.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <param name="source">A sequence that contains elements to be counted.</param>
+        /// <param name="count">
+        ///     When this method returns, contains the count of <paramref name="source" /> if successful,
+        ///     or -1 if the method failed to determine the count.</param>
+        /// <returns>
+        ///   <see langword="true" /> if the count of <paramref name="source"/> can be determined without enumeration;
+        ///   otherwise, <see langword="false" />.
+        /// </returns>
+        /// <remarks>
+        ///   The method performs a series of type tests, identifying common subtypes whose
+        ///   count can be determined without enumerating; this includes <see cref="ICollection{T}"/>,
+        ///   <see cref="ICollection"/> as well as internal types used in the LINQ implementation.
+        ///
+        ///   The method is a constant-time operation.
+        /// </remarks>
+        public static bool TryGetNonEnumeratingCount<TSource>(this IEnumerable<TSource> source, out int count)
```

Comment:
Can we delete https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs#L21 and use this instead?

Reply:
I thought about it, but it's slightly different in that it doesn't do null checks or test for `ICollection`. So perhaps it makes sense to keep a separate method for internals?

Reply:
I don't have a strong opinion, whatever you decide.  But in the context of all the work callers of this method are doing, and even the interface checks this method is doing, I don't think a null check will show up as even a blip, and if an ICollection check is valuable here, presumably it would also be valuable in the existing call sites?  I'm just speculating.  Up to you.  I always prefer to avoid internal variants of public methods because it then suggests that if the public method isn't good enough for us, why do we expect it to be good enough for anyone else?

Reply:
Sounds reasonable, I was making the assumption that `ICollection` was omitted intensionally in the internal method unlike `Enumerable.Count`. While we're at it, do you think it might make sense to also test for `IReadOnlyCollection<T>`? I couldn't find any implementations in BCL that don't also implement `ICollection<T>`, but that might not be the case in 3rd party types. I think the expectation from callers of this method is to make the best effort at deducing the count.

Reply:
I think we should use or not use IROC consistently. Not include it in this change,  and then have an end to end evaluation of whether it should be used throughout LINQ. In the past,  we decided no.  Some aspects of the equation have changed,  so we can look again.

Reply:
> if the public method isn't good enough for us, why do we expect it to be good enough for anyone else?

Aren't we using NRT internally? so the null checks are just redundant there. 
(I do see you addressed that in your comment as insignificant, perf-wise)

Reply:
> so the null checks are just redundant there

Just because NRT is being used,  it doesn't mean runtime checks are redundant. 

But even if they were, there are tons of redundant null checks all over the place.  These are not going to matter.